lucene/core/src/java/org/apache/lucene/index/ExitableDirectoryReader.java

```diff
@@ -100,13 +109,150 @@ public CacheHelper getCoreCacheHelper() {
 
   }
 
+  /**
+   * Wrapper class for another PointValues implementation that is used by ExitableFields.
+   */
+  public static class ExitablePointValues extends PointValues {
+
+    private final PointValues in;
+    private final QueryTimeout queryTimeout;
+
+    public ExitablePointValues(PointValues in, QueryTimeout queryTimeout) {
+      this.in = in;
+      this.queryTimeout = queryTimeout;
+      checkAndThrow();
+    }
+
+    /**
+     * Throws {@link ExitingReaderException} if {@link QueryTimeout#shouldExit()} returns true,
+     * or if {@link Thread#interrupted()} returns true.
+     */
+    private void checkAndThrow() {
+      if (queryTimeout.shouldExit()) {
+        throw new ExitingReaderException("The request took too long to iterate over point values. Timeout: "
+            + queryTimeout.toString()
+            + ", PointValues=" + in
+        );
+      } else if (Thread.interrupted()) {
+        throw new ExitingReaderException("Interrupted while iterating over point values. PointValues=" + in);
+      }
+    }
+
+    @Override
+    public void intersect(IntersectVisitor visitor) throws IOException {
+      checkAndThrow();
+      in.intersect(new ExitableIntersectVisitor(visitor, queryTimeout));
+    }
+
+    @Override
+    public long estimatePointCount(IntersectVisitor visitor) {
+      checkAndThrow();
+      return in.estimatePointCount(visitor);
+    }
+
+    @Override
+    public byte[] getMinPackedValue() throws IOException {
+      checkAndThrow();
+      return in.getMinPackedValue();
+    }
+
+    @Override
+    public byte[] getMaxPackedValue() throws IOException {
+      checkAndThrow();
+      return in.getMaxPackedValue();
+    }
+
+    @Override
+    public int getNumDataDimensions() throws IOException {
+      checkAndThrow();
+      return in.getNumDataDimensions();
+    }
+
+    @Override
+    public int getNumIndexDimensions() throws IOException {
+      checkAndThrow();
+      return in.getNumIndexDimensions();
+    }
+
+    @Override
+    public int getBytesPerDimension() throws IOException {
+      checkAndThrow();
+      return in.getBytesPerDimension();
+    }
+
+    @Override
+    public long size() {
+      checkAndThrow();
+      return in.size();
+    }
+
+    @Override
+    public int getDocCount() {
+      checkAndThrow();
+      return in.getDocCount();
+    }
+  }
+
+  public static class ExitableIntersectVisitor implements PointValues.IntersectVisitor {
+
+    public static final int MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK = 10;
+
+    private final PointValues.IntersectVisitor in;
+    private final QueryTimeout queryTimeout;
+    private int calls = 0;
```

Comment:
Minor: you don't need the `= 0` -- java does that for you by default.

Reply:
Yes, removed in 11fd3de51c31ef4c0e2a660ce22df018d194ce7e, thank you.